File: 560-subarray-sum-equals-k/subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        //prefix sums go into hashmap
        int n = nums.length;
        Map<Integer, Integer> sumToCount = new HashMap<>();
        sumToCount.put(0, 1);
        int res = 0;
        int currentSum = 0;
        for(int i=0; i<n; i++){
            //calculate current prefix sum
            //if k - current prefix sum exists in map, add to result
            // add in map current prefix sum
            currentSum += nums[i];
            int target = currentSum-k;
            if(sumToCount.containsKey(target)){
                res += sumToCount.get(target);
            }
            sumToCount.put(
                currentSum,
                sumToCount.getOrDefault(currentSum, 0)+1
            );
        }
        return res;

        

    }
}

File: subarray-sum-equals-k.cpp
class Solution {
    public int subarraySum(int[] nums, int k) {
        //work with prefix sums
        //at each prefix sum, check if the value of prefix sum - k was already a prefix sum,
        // and if yes, how many times
    
        Map<Integer, Integer> prefixSumCounts = new HashMap<>();
        prefixSumCounts.put(0,1);
        int res = 0;
        int currentSum = 0;
        for (int num : nums){
            currentSum +=num;
            if(prefixSumCounts.containsKey(currentSum-k)){
                res += prefixSumCounts.get(currentSum-k);
            }
            prefixSumCounts.put(currentSum, prefixSumCounts.getOrDefault(currentSum, 0)+1);

        }
        return res;
    }
}